### ISSUE TITLE
Access-Control-Allow-Credential should be set to false for /sockjs

### ISSUE THREAD

**Issue by mitar** [meteor/meteor#6389]
I was looking into fast render and how they get login token to the server side. I noticed that they use cookies and the discussion about [cookie security here](https://github.com/kadirahq/fast-render#security).

I do not agree completely with that analysis and I think most issues with cookies would be resolved if the cookie origin (site hosting the Meteor app) would ever set `Access-Control-Allow-Credential` to `true` (then it does not matter if cookie is used or not, or send or not, because browsers will not provide content back to the malicious site doing cross-origin request trying to misuse the cookie).

To my surprise, sock.js is setting `Access-Control-Allow-Credential` to `true` by default on the `/sockjs`. I guess because in most cases sites use cookies to authenticate `/sockjs` connection. But Meteor is not like that. So I think Meteor `/sockjs` should make sure that `Access-Control-Allow-Credential` is set to `false`.

See this ticket for more information: https://github.com/kadirahq/fast-render/issues/162

And see this [mailing list post](https://groups.google.com/forum/#!msg/meteor-talk/1Fg4rNk9JZM/ELX3672QsrEJ) for some information how `/sockjs` could be misused.

**Comment by zol** [comment]
Hey Mitar, Thanks for continuing this discussion. I'm going to leave this open for now to see what others think - I don't know enough about the implications of `Access-Control-Allow-Credential` to `false`.

**Comment by mitar** [comment]
It just does not allow any cookies to be send for cross-origin requests. And because Meteor is not using any cookies on `/sockjs` it should just improve the security.

**Comment by kachkaev** [comment]
I'm not sure if this is related, but I probably get an issue with [`mapbox-gl`](https://github.com/mapbox/mapbox-gl-js) because `Access-Control-Allow-Credential` is `true` in Meteor.

```txt
XMLHttpRequest cannot load https://api.mapbox.com/styles/v1/mapbox/light-v8?access_token=pk.eyJ1IjXXXXXXX. A wildcard '*' cannot be used in the 'Access-Control-Allow-Origin' header when the credentials flag is true. Origin 'http://42.42.42.42:3000' is therefore not allowed access. The credentials mode of an XMLHttpRequest is controlled by the withCredentials attribute.
```

This exception is thrown immediately when a meteor App is opened in the browser. My guess is that Mapbox GL is trying to init a socket to start fetching vector tiles, but Chrome does not allow that.

@mitar, do you think this can be related? Any idea on how to overcome the error?

**Comment by mitar** [comment]
@zol: I have learned even more later on. That [sock.js explicitly ignores cookies and other authorization headers](https://github.com/sockjs/sockjs-node#various-issues-and-design-considerations) as part of their security model. There is really no reason to then have `Access-Control-Allow-Credential` set to `true` for `/sockjs` path.

**Comment by mitar** [comment]
@kachkaev: You can probably set `Origin` explicitly, something like:

```
var rootUrl = __meteor_runtime_config__.ROOT_URL;
BrowserPolicy.content.allowConnectOrigin(rootUrl);
BrowserPolicy.content.allowConnectOrigin(rootUrl.replace('http', 'ws'));
```

See [here for more information](https://meteorhacks.com/xss-and-meteor/).

**Comment by kachkaev** [comment]
Thanks @mitar!

Seems like this helped a bit, but not completely:

```
svg.react.js:61Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' 'unsafe-inline'".
```

Never mind. I'll try [mxstbr/react-boilerplate](https://github.com/mxstbr/react-boilerplate) for that MapboxGL project and keep an eye on how things evolve in Meteor. Will be missing Meteor methods and live data sync out of box for now =)

**Comment by mitar** [comment]
@kachkaev: This is really not a forum for help with Meteor. The issues you are having is simply a question of configuration. Have you read the link I gave you above for more information? In the same way you can configure which origin can connect, you can also configure to use unsafe-eval.

[One more link with info](https://themeteorchef.com/snippets/using-the-browser-policy-package/).

**Comment by hwillson** [comment]
To help provide a more clear separation between feature requests and bugs, and to help clean up the feature request backlog, Meteor feature requests are now being managed under the https://github.com/meteor/meteor-feature-requests repository. 

This feature request will be closed here, but anyone interested in migrating this feature request to the new repository (to make sure it stays active), can click [here](https://github.com/meteor/meteor-feature-requests/issues/new?title=Access-Control-Allow-Credential%20should%20be%20set%20to%20false%20for%20%2Fsockjs&body=Migrated%20from:%20meteor/meteor%236389) to start the feature request migration process. This manual migration process is intended to help identify which of the older feature requests are still considered to be of value to the community. Thanks!

**Comment by mitar** [comment]
I think this is a bug or even a security issue, not necessary a feature request.

**Comment by hwillson** [comment]
Fair point @mitar - I was on the fence on this one as well. Okay, I'll re-open it here and mark it as `pull-requests-encouraged`. If anyone is interested in working on this, please test thoroughly to verify there are no initially hidden implications to making this change. Thanks!
